### PR TITLE
[Sampler] Stop upcasting logits to fp32 in apply_sampling_params

### PR DESCRIPTION
**Pull request by mgoin**
## Summary

`apply_sampling_params` copies the logits into a new fp32 tensor before applying sampling params. Generation models emit bf16 logits by default (except in RL), so this copy is a pure upcast that costs `num_logits * vocab * 4` bytes. Under spec decode there are `num_reqs * (1 + num_spec_tokens)` logits instead of `num_reqs`, so the copy reaches multiple GiB. On Qwen3-8B + a dflash speculator it tries to allocate 4.12 GiB and OOMs during warmup.

Every op in the sample chain already upcasts to fp32 on load except `apply_top_k_top_p_triton`, which asserted fp32 input tensor. Upcasting its logits loads makes the whole chain dtype-preserving, so the copy can go entirely. It's now only taken when `raw_logprobs` needs the original logits intact.

Supersedes #48630 and #48037, which fixed the same OOM by *accounting* for the buffer rather than removing it (#48037 gives up ~4 GiB of KV cache to do so).

## Memory

Repro on H100: `vllm serve Qwen/Qwen3-8B --spec-model RedHatAI/Qwen3-8B-speculator.dflash
--spec-method dflash --spec-tokens 7 --speculative-config '{"attention_backend":"FLASHINFER"}'`

|  | peak activation | KV cache | result |
| --- | --- | --- | --- |
| main | 3.04 GiB | 52.02 GiB | **OOM** (tried to allocate 4.12 GiB) |
| this PR | 3.04 GiB | 52.02 GiB | boots |

## Correctness

GSM8K, 3 runs, Qwen3-8B + dflash speculator:

| sampling params | main | this PR |
| --- | --- | --- |
| `top_k=20, top_p=0.95, temp=0.6` (Qwen3 defaults) | 0.890 / 0.892 / 0.892 | 0.894 / 0.894 / 0.894 |
| `top_k=0, top_p=0.95, temp=0.6` (pure top-p) | 0.896 / 0.896 / 0.896 | 0.894 / 0.896 / 0.896 |

GPQA, gpt-oss-20b, TP=2 (non-spec path, `temp=1.0`): 0.5650 on main, 0.5827 on PR

## Testing

Removing the fp32 copy meant it was silently normalizing two things beyond dtype — **contiguity** and freshness — so I went looking for what relied on that. Two real bugs turned up, both now fixed here with a regression test (`test_flashinfer_sample_padded_vocab`):

- **FlashInfer + padded vocab crashed.** `compute_logits` slices off the vocab padding, so for a non-64-aligned vocab (opt, gpt2, gpt-j) the logits are a strided view. FlashInfer asserts contiguous fp32, so `top_k` + `top_p` aborted once batch size > 1. A single row stays contiguous despite the padded stride, which is why it hid at batch 1. Fixed with `logits.float().contiguous()` at the FlashInfer boundary — matching the guard the V1 sampler already has.
- **top-p PyTorch fallback undershot the nucleus.** It softmaxed + cumsummed over the whole vocab in the input dtype; in bf16 the accumulation lost enough precision that `p=0.95` kept ~0.947. This is the `batch < 8` path (low-concurrency decode). Fixed by forcing fp32, matching `flashinfer_sample`.

What I checked to gain confidence:

- **Every logits kernel** (penalties, temperature, min_p, bad_words, logit_bias, gumbel, logprobs, top-k/top-p) run against a strided view vs. its contiguous copy — all bit-identical.
- **opt-125m end-to-end** (padded vocab → strided; fp16 model): greedy, greedy + full penalty/bias/min-tokens chain, seeded top-k/top-p, and raw logprobs all match `main` exactly.
- **FlashInfer sampler path** — it's on by default and bypasses the triton kernel entirely for non-spec models, so it needed its own coverage. GSM8K unseeded, 8×500 questions: `0.895` main vs `0.896` PR (0.2σ; unseeded run-to-run spread alone is ~3.8 points).

One user-visible change worth calling out: `processed_logprobs` now derives from bf16 rather than fp32, so returned logprobs shift by ≤0.01. Token choices are unaffected, and `head_dtype=float32` (the RL path) stays fp32.

AI assistance was used for this change.

### CODE REVIEW

**Comment by mgoin** [comment]
Need to test what the FlashInfer sampler is okay with
EDIT: Seems like no issues with flashinfer sampler

**Comment by mgoin** [comment]
Hey @cakeng would you have any thoughts on this or concerns about possible changes in behavior?

File: vllm/v1/sample/ops/topk_topp_triton.py
Comment (cakeng) [inline vllm/v1/sample/ops/topk_topp_triton.py:853]: Hi @mgoin, I don't think the behavior will change for the triton sampler as long as the buffer/tables are fp32 and all logits are casted to fp32 on load. One question though, is the downcast to fp16/bf16 automatic on the final store? (line 853)

Comment (mgoin) [reply vllm/v1/sample/ops/topk_topp_triton.py:853]: Yes the downcast is automatic!

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py [hunk 1]
```diff
@@ -110,13 +110,18 @@ def __call__(
 
         draft_sampled = input_batch.input_ids[input_batch.logits_indices]
         pos = input_batch.positions[input_batch.logits_indices]
+        return_logprobs = (
+            self.sampler.sampling_states.max_num_logprobs(input_batch.idx_mapping_np)
+            != NO_LOGPROBS
+        )
```
Comment (njhill) [inline vllm/v1/worker/gpu/spec_decode/rejection_sampler.py]: Should this also take into account possibly-provided specific logprobs token ids set, i.e. per https://github.com/vllm-project/vllm/pull/48641/changes#diff-2b1b7728d87b07fc37d8428df43f5fdf0f8b45229d0d6561357613bd9d9120f6R88-R92?

If so should we stash this common predicate in InputBatch and use it in both places? I wonder if it would be cleaner to pass something like `return_logprobs` to `apply_sampling_params` (with maybe more appropriate name for that context) instead of `in_place`, and then the `needs_raw_logits` logic can just be inlined there?

**Review by njhill — COMMENTED** [review]
Thanks @mgoin this looks great!

The last thing holding up switching over to MRV2 for remaining (non-pooling) models is some OOMs like this, so this should hopefully help with that https://buildkite.com/vllm/ci/builds/78392/list?jid=019f6ae5-772c-4e05-b5e3-19f281f7fd37&tab=output

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py (hunk 1 above)
Comment (mgoin) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler.py]: Great suggestion! I liked consolidating it into `Sampler.returns_logprobs()`, moving the copy decision inside `apply_sampling_params` so `RejectionSampler` no longer carries its own version. The small cost we pay for the simplicity is `Sampler.__call__` recomputes `max_num_logprobs`/`max_per_req_token_ids` in the rare spec decode `if return_logprobs:` block.

InputBatch only holds batch layout (`idx_mapping_np`), while the predicate reads sampler-owned state so it makes most sense belonging on Sampler

Comment (njhill) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler.py]: Thanks @mgoin!

> The small cost we pay for the simplicity is `Sampler.__call__` recomputes `max_num_logprobs`/`max_per_req_token_ids` in the rare spec decode `if return_logprobs:` block.

I guess we are now also computing those twice in the non-spec decode `return_logprobs` path too now, and so I think could be doing that same indexing of the two numpy arrays up to 3 times per step? But agree this is much less common case. We might still want to think a bit more how we can cleanly/clearly avoid that redundancy but let's worry about it as a follow-on.

**Review by njhill — APPROVED** [review]

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py (hunk 1 above)
Comment (mgoin) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler.py]: Doh you are right.. I'll start working on the shape of fixing this as a follow up

File: vllm/v1/worker/gpu/sample/sampler.py [hunk 2]
```diff
@@ -153,8 +160,13 @@ def apply_sampling_params(
         expanded_local_pos: torch.Tensor,
         skip_top_k_top_p: bool = False,
     ) -> torch.Tensor:
-        # Copy logits to a new FP32 tensor.
-        logits = torch.empty_like(logits, dtype=torch.float32).copy_(logits)
+        # The ops below upcast to fp32 internally, so the input dtype is kept and
+        # mutated in place. Only raw_logprobs reads the unmodified logits
+        # afterward, so copy just for that case.
+        if self.logprobs_mode == "raw_logprobs" and self.returns_logprobs(
```
Comment (njhill) [inline vllm/v1/worker/gpu/sample/sampler.py]: This should be changed now that https://github.com/vllm-project/vllm/pull/47680 has been merged.
```suggestion
        if self.logprobs_mode.startswith("raw_") and self.returns_logprobs(
```
